src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp

```diff
@@ -0,0 +1,411 @@
+#include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+PartMovesBetweenShardsOrchestrator::PartMovesBetweenShardsOrchestrator(StorageReplicatedMergeTree & storage_)
+    : storage(storage_)
+    , zookeeper_path(storage.zookeeper_path)
+    , logger_name(storage.getStorageID().getFullTableName() + " (PartMovesBetweenShardsOrchestrator)")
+    , log(&Poco::Logger::get(logger_name))
+    , entries_znode_path(zookeeper_path + "/part_moves_shard")
+{
+    task = storage.global_context.getSchedulePool().createTask(logger_name, [this]{ run(); });
+}
+
+void PartMovesBetweenShardsOrchestrator::run()
+{
+    if (need_stop)
+        return;
+
+    auto sleep_ms = 10;
+
+    try
+    {
+        sync();
+
+        if (step())
+            sync();
+        else
+            sleep_ms = 3 * 1000;
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, __PRETTY_FUNCTION__);
+    }
+
+
+    task->scheduleAfter(sleep_ms);
+}
+
+void PartMovesBetweenShardsOrchestrator::shutdown()
+{
+    need_stop = true;
+    task->deactivate();
+    LOG_TRACE(log, "PartMovesBetweenShardsOrchestrator thread finished");
+}
+
+void PartMovesBetweenShardsOrchestrator::sync()
+{
+    std::lock_guard lock(state_mutex);
+
+    entries.clear();
+
+    auto zk = storage.getZooKeeper();
+
+    Strings task_names = zk->getChildren(entries_znode_path);
+    for (auto const & task_name : task_names)
+    {
+        PartMovesBetweenShardsOrchestrator::Entry e;
+        Coordination::Stat stat;
+
+        e.znode_path = entries_znode_path + "/" + task_name;
+
+        auto entry_str = zk->get(e.znode_path, &stat);
+        e.fromString(entry_str);
+
+        e.version = stat.version;
+        e.znode_name = task_name;
+
+        entries[task_name] = std::move(e);
+    }
+}
+
+bool PartMovesBetweenShardsOrchestrator::step()
+{
+    if (!storage.is_leader)
+        return false;
+
+    auto zk = storage.getZooKeeper();
+
+    std::optional<Entry> entry_to_process;
+
+    /// Try find an entry to process and copy it.
+    {
+        std::lock_guard lock(state_mutex);
+
+        for (auto const & entry : entries | boost::adaptors::map_values)
+        {
+            if (entry.state.value == EntryState::DONE || entry.state.value == EntryState::CANCELLED)
+                continue;
+
+            entry_to_process.emplace(entry);
+            break;
+        }
+    }
+
+    if (!entry_to_process.has_value())
+        return false;
+
+    try
+    {
+
+        /// Since some state transitions are long running (waiting on replicas acknowledgement we create this lock to avoid
+        /// other replicas trying to do the same work. All state transitions should be idempotent so is is safe to lose the
+        /// lock and have another replica retry.
+        ///
+        /// Note: This blocks all other entries from being executed. Technical debt.
+        auto entry_node_holder = zkutil::EphemeralNodeHolder::create(entry_to_process->znode_path + "/lock_holder", *zk, storage.replica_name);
```

Comment:
Correct. This should be held until function exit.

Additionally it seems that a makeCheckRequest is needed so that other replicas won't issue tasks (SYNC, FETCH, DELETE) after another replica transitioned to a different state. There must be a reason I haven't done this before...